0002-ether-htlc-dynamic-final-addresses.md

```diff
@@ -0,0 +1,121 @@
+# Ether HTLC with dynamic final addresses
+
+* Authors: Lloyd Fournier
+* Date: 2019-04-04
+
+Issue: [#40](https://github.com/comit-network/RFCs/issues/40)
+
+## Context
+
+The RFC003 [Bitcoin HTLC](https://github.com/comit-network/RFCs/blob/master/RFC-005-SWAP-Basic-Bitcoin.md) and [Ether HTLC](https://github.com/comit-network/RFCs/blob/master/RFC-007-SWAP-Basic-Ether.md) have different functionality.
+Bitcoin's UTXO model means you specify which identity will finally own the funds in the redeem and refund transactions.
+In the Ethereum HTLC this is embedded in the contract from the beginning.
+
+Both contracts are representative of the paradigms of each blockchain but their differences get in the way writing generic code for interacting with them.
+The nature of Bitcoin allows the COMIT node to handle all the keys involved in the HTLC and simply ask the user for where they want the funds when the time comes.
+To me this is good design and what we want in COMIT even outside of atomic swap protocols:
+
+1. The users negotiate a protocol with no keys required from the users' wallets
+2. The protocol execution starts with users funding the initial contracts from their wallets
+3. The COMIT nodes execute the protocol (via the user's instruction and/or automatically).
+4. The protocol execution completes by the users' getting the final assets they own into their wallets.
+
+As a rule of thumb, the COMIT node should own the private keys of whatever identities are embedded in the contracts on the blockchains.
+
+Since Ethereum is flexible we can restructure it to mimic how Bitcoin works and fit it into the above flow.
+
+## Research
+
+Bitcoin requires you witness the HTLC output with the secret and a signature (or a just the refund signature after a timeout).
+The transaction that has the witness will specify where to redeem the funds in its output.
+
+To follow the same pattern in Ethereum, our redeem transaction will need in its data:
+
+1. The secret
+2. The address to give the money to (the final redeem address)
+3. a signature under the redeem identity from the COMIT node
+
+The refund would be the same without (1)
+
+### The Final Address
+
+The decision we have to make here is whether the final address is:
+
+1. Provided in the data
+2. The sender address
+
+(2) is nice because it means that there is no way to mess it up and send a transaction to an address you don't own. It also saves a bit of gas.
+
+On the other hand, that might be a feature.
+You may want to send the asset in the HTLC to a contract or something, so I'm leaning towards (1).
```

Comment:
I am more keen on (2) for simplicity/cost/security (as you mention) reasons and change to (1) if the feature is requested.

Reply:
I think if we don't want to change our core concepts then (1) is the way to go: 

* If we go for (2) the comit-node would need funds to send a transaction from the intermediate account (if we want to use the same approach as for Bitcoin).
* You mentioned we want to mimic Bitcoin

Reply:
I would vote for option 1 mainly because of the reasons Philipp mentioned.

In addition, I see the following problem:

Sending the money to the sender's address requires the user to choose, which address the want to send it from when fetching the action. As you mentioned further down, the final redeem address needs to be signed for security reasons.

What if the user sends it from a different address though? This would mean the redeeming fails and debugging that requires serious knowledge about how the contracts work. To avoid this problem altogether, we should put the final redeem address into the `data`. This allows the redeeming to happen from any address and not just the address the user wants the money on.

I think (2) can cause way too many problems here and the overhead of (1) seems to be negligible (only the cost of the added bytes in the TX I presume).